fix: enable native streaming for Azure o4 models

colesmcintosh opened the pull request:
## Title

Enable native streaming for Azure o4 models

## Relevant issues

Fixes #12150

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🐛 Bug Fix

## Changes

- Updated `should_fake_stream` method in `litellm/llms/azure/chat/o_series_transformation.py` to return `False` for o4 models (including o4-mini)
- Azure o4 models now support native streaming per [Microsoft documentation](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/reasoning?tabs=python-secure%2Cpy#api--feature-support)
- Added tests to verify o4 streaming behavior in `tests/llm_translation/test_azure_o_series.py`

### Test Results

Tests pass successfully:
```
tests/llm_translation/test_azure_o_series.py::test_should_fake_stream_for_o4_models PASSED
tests/llm_translation/test_azure_o_series.py::test_azure_o3_streaming PASSED
tests/llm_translation/test_azure_o_series.py::test_azure_o4_streaming PASSED
```

### Changes Summary

1. Modified `should_fake_stream` to check for both "o3" and "o4" in model names
2. Updated docstring to reflect that both O3 and O4 support native streaming
3. Added comprehensive tests for o4 streaming behavior

colesmcintosh committed:
fix: enable native streaming for Azure o4 models

- Update should_fake_stream to return False for o4 models (o4-mini, o4)
- Azure o4 models now support native streaming per Microsoft documentation
- Add tests to verify o4 streaming behavior

Fixes #12150

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/azure-o4-streaming

krrishdholakia commented on tests/llm_translation/test_azure_o_series.py:
```diff
@@ -127,6 +127,53 @@ def test_azure_o3_streaming():
         assert "stream" in mock_create.call_args.kwargs
 
 
+def test_azure_o4_streaming():
```
can we move mock tests in `test_litellm/` so we can confirm they pass?

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/azure-o4-streaming

colesmcintosh committed:
test: move o4 streaming mock tests to test_litellm directory

- Move test_azure_o4_streaming and test_should_fake_stream_for_o4_models from tests/llm_translation/ to tests/test_litellm/llms/azure/chat/
- Create new file test_azure_o4_streaming.py in proper test location per PR feedback
- Tests pass in new location

ishaan-jaff commented on litellm/llms/azure/chat/o_series_transformation.py:
```diff
@@ -78,15 +78,15 @@ def should_fake_stream(
-        if (
-            model and "o3" in model
-        ):  # o3 models support streaming - https://github.com/BerriAI/litellm/issues/8274
+        if model and ("o3" in model or "o4" in model):
```
instead use a helper method called does_o_series_support_native_streaming()

ishaan-jaff reviewed (changes requested):
reviewed

colesmcintosh committed:
refactor: create helper method for O-series streaming support check

- Add does_o_series_support_native_streaming() helper in azure/common_utils.py
- Update should_fake_stream() to use the new helper method
- Apply Black formatting to modified files
- Tests continue to pass for o3/o4 streaming functionality

ishaan-jaff reviewed (changes requested):
-

colesmcintosh committed:
refactor: move does_o_series_support_native_streaming to AzureOpenAIO1Config

- Remove does_o_series_support_native_streaming from common_utils.py
- Implement does_o_series_support_native_streaming as a method in AzureOpenAIO1Config
- Update should_fake_stream() to call the new method
- Ensure functionality remains intact for O-series models

colesmcintosh committed:
revert: ensure common_utils.py and test_azure_o_series.py match main exactly

colesmcintosh committed:
fix(tests): clean up whitespace in test_azure_o4_streaming.py

- Remove unnecessary blank lines in the test_should_fake_stream_for_o4_models function.
- Ensure consistent formatting and maintain readability of the test code.

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/azure-o4-streaming
